src/Hl7.Fhir.Core/Serialization/ResourceReader.cs

```diff
@@ -41,11 +41,11 @@ public Resource Deserialize(Resource existing=null)
         {
             // If there's no a priori knowledge of the type of Resource we will encounter,
             // we'll have to determine from the data itself. 
-            var resourceTypeName = _reader.GetResourceTypeIndicator();
-            var mapping = _inspector.FindClassMappingForResource(resourceTypeName);
+
+            var mapping = _inspector.FindClassMappingForResource(_reader.InstanceType);
```

Comment:
Just like in ComplexTypeReader, there's an assumption here that InstanceType != null. I am not so sure...